### ISSUE TITLE
@ConditionalOnProperty does not have the same result in versions 1.5.* and 2.0.*

### ISSUE THREAD

**Issue by hengyunabc** [spring-projects/spring-boot#13449]
`@ConditionalOnProperty` does not have the same result in versions 1.5.* and 2.0.* 

## Code

[ConditionalOnProperty-demo.zip](https://github.com/spring-projects/spring-boot/files/2093307/ConditionalOnProperty-demo.zip)

`application.properties`:

```
test.fooKey=foo
test.bar-key=bar
```

```java
	@Bean
	@ConditionalOnProperty(name = "test.fooKey")
	public static String fooKeyNoDash() {
		return "@ConditionalOnProperty(name = \"test.fooKey\")";
	}

	@Bean
	@ConditionalOnProperty(name = "test.foo-key")
	public static String fooKeyDash() {
		return "@ConditionalOnProperty(name = \"test.foo-key\")";
	}

	@Bean
	@ConditionalOnProperty(name = "test.barKey")
	public static String barKeyNoDash() {
		return "@ConditionalOnProperty(name = \"test.barKey\")";
	}

	@Bean
	@ConditionalOnProperty(name = "test.bar-key")
	public static String barKeyDash() {
		return "@ConditionalOnProperty(name = \"test.bar-key\")";
	}
```

###  spring boot 1.5.13.RELEASE

```
application.properties: 

test.fooKey=foo
test.bar-key=bar

True  :@ConditionalOnProperty(name = "test.fooKey")
False :@ConditionalOnProperty(name = "test.foo-key")
True  :@ConditionalOnProperty(name = "test.barKey")
True  :@ConditionalOnProperty(name = "test.bar-key")
```

* `@ConditionalOnProperty(name = "test.barKey")` is True, because `RelaxedPropertyResolver` check all `RelaxedNames` of `test.barKey` .

### spring boot 2.0.2.RELEASE

```
application.properties: 

test.fooKey=foo
test.bar-key=bar

True  :@ConditionalOnProperty(name = "test.fooKey")
True  :@ConditionalOnProperty(name = "test.foo-key")
False :@ConditionalOnProperty(name = "test.barKey")
True  :@ConditionalOnProperty(name = "test.bar-key")
```

* `@ConditionalOnProperty(name = "test.foo-key")` is True, because `org.springframework.boot.context.properties.source.ConfigurationPropertyName`

**Comment by snicoll** [comment]
The Javadoc of `@ConditionalOnProperty` in 2.0 mentions that you must [use the canonical format](https://github.com/spring-projects/spring-boot/blob/ae35c17e8c6a31c98bf7427ecb87f4943982bc95/spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/condition/ConditionalOnProperty.java#L105). Both `test.fooKey` and `test.barKey` are invalid.

This is also mentioned [in the migration guide](https://github.com/spring-projects/spring-boot/wiki/Spring-Boot-2.0-Migration-Guide#relaxed-binding) and I just rephrased it slightly to make that a bit more explicit.

**Comment by hengyunabc** [comment]
@snicoll

I have a starter have to support both spring boot 1.5 and 2.0 .  
The users have two kinds config in their code: `test.fooKey=foo` and `test.foo-key=foo`.

To support 1.5 users, I need to write `@ConditionalOnProperty(name = "test.fooKey")`.
To support 2.0 users, I need to write `@ConditionalOnProperty(name = "test.foo-key")`.

Finally, I found the solution: 

```java
public class FooAnyNestedCondition extends AnyNestedCondition {
	public FooAnyNestedCondition() {
		super(ConfigurationPhase.REGISTER_BEAN);
	}
	@ConditionalOnProperty(name = "test.fooKey")
	static class FooKey {
	}
	@ConditionalOnProperty(name = "test.foo-key")
	static class FooKeyStash {
	}
}
```

The code is confusing...

---

I suggest that marking `@ConditionalOnProperty` have the same result in versions 1.5.* and 2.0.*

**Comment by snicoll** [comment]
@hengyunabc thanks for the additional feedback. We don't really recommend having an auto-config that works with bot `1.5.x` and `2.0.x` as quite a few things have changed (relaxed binding is one of them). Having said that, you have a point about the migration use case. I've flagged this one for team attention to check with the rest of the team.

**Comment by snicoll** [comment]
@hengyunabc on a second look, that's not how you should define a property in `1.5.x`. For relaxed binding to work properly there, you have to separate the prefix from the property, something like `@ConditionalOnProperty(prefix = "test", "name="foo-key")`. 

Also, the Javadoc of `@ConditionalOnProperty` requires you to use [the dashed notation](https://github.com/spring-projects/spring-boot/blob/6dc14af92dbfa7b5281072d9fdde241ce0da3679/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/condition/ConditionalOnProperty.java#L105). If you do this the behaviour is consistent in 1.5.x and 2.0.x.

What is new in 2.0.x is that you don't have to separate the prefix from the property for relaxed binding to work properly.

**Comment by hengyunabc** [comment]
@snicoll Thanks, I understand what you mean.  Before the 2.0 version appeared, I tried to support two styles of user config, so I wrote code like `@ConditionalOnProperty(name = "test.fooKey")`.

---

When using version 2.0.x , in my opinion, users should not have these burdens. For example, the `TestProperties` class:

```java
@ConfigurationProperties(prefix = "test")
public class TestProperties {
	String fooKey;
}
```

When using `@ConditionalOnProperty`， user have to convert to **the dashed notation**. 

```java
	@ConditionalOnProperty(name = "test.foo-key")
        @Bean
	public String someBean() {
		return null;
	}
```

For users who don’t know, it’s easy to ignore and make mistakes. I suggest that support `@ConditionalOnProperty(name = "test.fooKey")` will be better.
